File: core/src/main/java/io/github/msameer0/rhythmicrush/settings/SettingsManager.java
package io.github.msameer0.rhythmicrush.settings;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.utils.Json;
import com.badlogic.gdx.utils.JsonWriter;

/**
 * Manages the persistence, retrieval, and application of user-configurable game settings.
 * <p>
 * This class serves as a central hub for various preferences including audio levels,
 * graphical toggles (FPS caps, VSync), and UI visibility options. Settings are
 * serialized to and deserialized from a local JSON file using the LibGDX {@link Json}
 * utility to ensure configurations are saved across game sessions.
 * </p>
 */
public class SettingsManager {

    private static final String SAVE_PATH = "saves/settings.json";

    public boolean menuMusicEnabled = true;
    public float musicVolume = 1f;
    public boolean showHitboxes = false;
    public boolean showHitboxesOnDeath = false;
    public boolean lockCursorInGame = false;
    public boolean showFps = false;
    public boolean capFps = false;
    public int fpsCapValue = 60;
    public boolean enableVsync = false;
    public boolean showPercentage = true;
    public boolean showProgressBar = true;
    public boolean showAttempts = true;
    public boolean showBest = true;
    public float uiPadding = 12f;
    public float practiceButtonOpacity = 0.5f;

    /**
     * A data transfer object (DTO) used to represent a serializable snapshot of the settings.
     * This class is primarily used by {@link Json} for loading and saving settings to the local filesystem.
     */
    public static class Data {
        public boolean menuMusicEnabled = true;
        public float musicVolume = 1f;
        public boolean showHitboxes = false;
        public boolean showHitboxesOnDeath = false;
        public boolean lockCursorInGame = false;
        public boolean showFps = false;
        public boolean capFps = false;
        public int fpsCapValue = 60;
        public boolean enableVsync = false;
        public boolean showPercentage = true;
        public boolean showProgressBar = true;
        public boolean showAttempts = true;
        public boolean showBest = true;
        public float uiPadding = 12f;
        public float practiceButtonOpacity = 0.5f;
    }

    private final Json json;

    /**
     * Initializes a new SettingsManager, configures the JSON serializer,
     * and attempts to load existing settings from the local storage.
     */
    public SettingsManager() {
        json = new Json();
        json.setOutputType(JsonWriter.OutputType.json);
        json.setUsePrototypes(false);
        load();
    }

    /**
     * Persists the current settings to a local JSON file.
     * <p>
     * This method captures a snapshot of the current configuration states,
     * serializes them into a formatted JSON string, and writes the output
     * to the local storage path defined by {@link #SAVE_PATH}.
     * </p>
     */
    public void save() {
        Gdx.app.log("SettingsManager", "Saving settings...");
        try {
            Data snapshot = new Data();
            snapshot.menuMusicEnabled = menuMusicEnabled;
            snapshot.musicVolume = musicVolume;
            snapshot.showHitboxes = showHitboxes;
            snapshot.showHitboxesOnDeath = showHitboxesOnDeath;
            snapshot.lockCursorInGame = lockCursorInGame;
            snapshot.showFps = showFps;
            snapshot.capFps = capFps;
            snapshot.fpsCapValue = fpsCapValue;
            snapshot.enableVsync = enableVsync;
            snapshot.showPercentage = showPercentage;
            snapshot.showProgressBar = showProgressBar;
            snapshot.showAttempts = showAttempts;
            snapshot.showBest = showBest;
            snapshot.uiPadding = uiPadding;
            snapshot.practiceButtonOpacity = practiceButtonOpacity;
            FileHandle file = Gdx.files.local(SAVE_PATH);
            file.parent().mkdirs();
            file.writeString(json.prettyPrint(snapshot), false);
            Gdx.app.log("SettingsManager", "Settings saved successfully.");
        } catch (Exception e) {
            Gdx.app.error("SettingsManager", "Failed to save: " + e.getMessage());
        }
    }

    /**
     * Loads the settings from the local storage file.
     * <p>
     * This method attempts to read the JSON file at {@link #SAVE_PATH}. If the file exists,
     * it deserializes the content into a {@link Data} object and updates the current
     * instance's fields with the stored values. If the file is missing or corrupted,
     * the default settings are retained and an error is logged.
     * </p>
     */
    private void load() {
        Gdx.app.log("SettingsManager", "Loading settings...");
        try {
            FileHandle file = Gdx.files.local(SAVE_PATH);
            if (!file.exists()) {
                Gdx.app.log("SettingsManager", "No settings file found. Using defaults.");
                return;
            }
            Data d = json.fromJson(Data.class, file);
            if (d == null) return;
            menuMusicEnabled = d.menuMusicEnabled;
            musicVolume = d.musicVolume;
            showHitboxes = d.showHitboxes;
            showHitboxesOnDeath = d.showHitboxesOnDeath;
            lockCursorInGame = d.lockCursorInGame;
            showFps = d.showFps;
            capFps = d.capFps;
            fpsCapValue = d.fpsCapValue;
            enableVsync = d.enableVsync;
            showPercentage = d.showPercentage;
            showProgressBar = d.showProgressBar;
            showAttempts = d.showAttempts;
            showBest = d.showBest;
            uiPadding = d.uiPadding;
            practiceButtonOpacity = d.practiceButtonOpacity;
            Gdx.app.log("SettingsManager", "Settings loaded successfully.");
        } catch (Exception e) {
            Gdx.app.error("SettingsManager", "Failed to load: " + e.getMessage());
        }
    }

    /**
     * Applies the frames per second (FPS) limit to the game's graphics settings.
     * <p>
     * If {@link #capFps} is enabled, the foreground FPS is restricted to the value
     * defined in {@link #fpsCapValue}. Otherwise, the cap is set to 0, which
     * effectively disables the FPS limit in LibGDX.
     * </p>
     */
    public void applyFpsCap() {
        Gdx.graphics.setForegroundFPS(capFps ? fpsCapValue : 0);
    }

    /**
     * Applies the vertical synchronization (VSync) setting to the graphics configuration.
     * <p>
     * VSync is automatically enabled on non-desktop platforms to ensure rendering stability.
     * On desktop platforms, the state is determined by the {@link #enableVsync} toggle.
     * </p>
     */
    public void applyVsync() {
        boolean vsync = (Gdx.app.getType() != com.badlogic.gdx.Application.ApplicationType.Desktop)
            || enableVsync;
        Gdx.graphics.setVSync(vsync);
    }
}

File: core/src/main/kotlin/io/github/msameer0/rhythmicrush/settings/SettingsManager.kt
package io.github.msameer0.rhythmicrush.settings

import com.badlogic.gdx.Application
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.utils.Json
import com.badlogic.gdx.utils.JsonWriter

/**
 * Manages game settings, including persistence and application of graphics/audio configurations.
 */
class SettingsManager {
    var menuMusicEnabled: Boolean = true
    var musicVolume: Float = 1f
    var sfxVolume: Float = 1f
    var deathEffectEnabled: Boolean = true
    var showHitboxes: Boolean = false
    var showHitboxesOnDeath: Boolean = false
    var lockCursorInGame: Boolean = false
    var showFps: Boolean = false
    var capFps: Boolean = false
    var fpsCapValue: Int = 60
    var enableVsync: Boolean = false
    var showPercentage: Boolean = true
    var showProgressBar: Boolean = true
    var showAttempts: Boolean = true
    var showBest: Boolean = true
    var uiPadding: Float = 12f
    var practiceButtonOpacity: Float = 0.5f
    var pulseOrbs: Boolean = true

    /** Stores the serializable representation of game settings. */
    class Data {
        var menuMusicEnabled: Boolean = true
        var musicVolume: Float = 1f
        var sfxVolume: Float = 1f
        var deathEffectEnabled: Boolean = true
        var showHitboxes: Boolean = false
        var showHitboxesOnDeath: Boolean = false
        var lockCursorInGame: Boolean = false
        var showFps: Boolean = false
        var capFps: Boolean = false
        var fpsCapValue: Int = 60
        var enableVsync: Boolean = false
        var showPercentage: Boolean = true
        var showProgressBar: Boolean = true
        var showAttempts: Boolean = true
        var showBest: Boolean = true
        var uiPadding: Float = 12f
        var practiceButtonOpacity: Float = 0.5f
        var pulseOrbs: Boolean = true
    }

    private val json: Json = Json()

    init {
        json.setOutputType(JsonWriter.OutputType.json)
        json.setUsePrototypes(false)
        load()
    }

    fun save() {
        Gdx.app.log("SettingsManager", "Saving settings...")
        try {
            val snapshot = Data()
            snapshot.menuMusicEnabled = menuMusicEnabled
            snapshot.musicVolume = musicVolume
            snapshot.sfxVolume = sfxVolume
            snapshot.deathEffectEnabled = deathEffectEnabled
            snapshot.showHitboxes = showHitboxes
            snapshot.showHitboxesOnDeath = showHitboxesOnDeath
            snapshot.lockCursorInGame = lockCursorInGame
            snapshot.showFps = showFps
            snapshot.capFps = capFps
            snapshot.fpsCapValue = fpsCapValue
            snapshot.enableVsync = enableVsync
            snapshot.showPercentage = showPercentage
            snapshot.showProgressBar = showProgressBar
            snapshot.showAttempts = showAttempts
            snapshot.showBest = showBest
            snapshot.uiPadding = uiPadding
            snapshot.practiceButtonOpacity = practiceButtonOpacity
            snapshot.pulseOrbs = pulseOrbs
            val file = Gdx.files.local(SAVE_PATH)
            file.parent().mkdirs()
            file.writeString(json.prettyPrint(snapshot), false)
            Gdx.app.log("SettingsManager", "Settings saved successfully.")
        } catch (exception: Exception) {
            Gdx.app.error("SettingsManager", "Failed to save: ${exception.message}")
        }
    }

    private fun load() {
        Gdx.app.log("SettingsManager", "Loading settings...")
        try {
            val file = Gdx.files.local(SAVE_PATH)
            if (!file.exists()) {
                Gdx.app.log("SettingsManager", "No settings file found. Using defaults.")
                return
            }
            val savedSettings = json.fromJson<Data?>(Data::class.java, file) ?: return
            menuMusicEnabled = savedSettings.menuMusicEnabled
            musicVolume = savedSettings.musicVolume
            sfxVolume = savedSettings.sfxVolume
            deathEffectEnabled = savedSettings.deathEffectEnabled
            showHitboxes = savedSettings.showHitboxes
            showHitboxesOnDeath = savedSettings.showHitboxesOnDeath
            lockCursorInGame = savedSettings.lockCursorInGame
            showFps = savedSettings.showFps
            capFps = savedSettings.capFps
            fpsCapValue = savedSettings.fpsCapValue
            enableVsync = savedSettings.enableVsync
            showPercentage = savedSettings.showPercentage
            showProgressBar = savedSettings.showProgressBar
            showAttempts = savedSettings.showAttempts
            showBest = savedSettings.showBest
            uiPadding = savedSettings.uiPadding
            practiceButtonOpacity = savedSettings.practiceButtonOpacity
            pulseOrbs = savedSettings.pulseOrbs
            Gdx.app.log("SettingsManager", "Settings loaded successfully.")
        } catch (exception: Exception) {
            Gdx.app.error("SettingsManager", "Failed to load: ${exception.message}")
        }
    }

    fun applyFpsCap() {
        Gdx.graphics.setForegroundFPS(if (capFps) fpsCapValue else 0)
    }

    fun applyVsync() {
        val vsync = (Gdx.app.type != Application.ApplicationType.Desktop)
            || enableVsync
        Gdx.graphics.setVSync(vsync)
    }

    companion object {
        private const val SAVE_PATH = "saves/settings.json"
    }
}
